File: src/main/java/com/yui/models/Artwork.java
package com.yui.models;

import java.util.Date;
import java.util.List;

import jakarta.persistence.Entity;
import jakarta.persistence.GeneratedValue;
import jakarta.persistence.GenerationType;
import jakarta.persistence.Id;
import jakarta.persistence.JoinColumn;
import jakarta.persistence.ManyToOne;
import jakarta.persistence.OneToMany;
import jakarta.persistence.Table;
import lombok.Data;

@Data
@Entity
@Table(name="artworks")
public class Artwork {
    @Id
    @GeneratedValue(strategy=GenerationType.AUTO)
    private int id;
    private String title;
    private String description;
    private String url;
    private double cryptoPrice;
    private double convertedPrice;
    private String currency;
    private boolean published;
    private Date createdAt;

    @ManyToOne
    @JoinColumn(name="user_id")
    private User user;

    @ManyToOne
    @JoinColumn(name="category_id")
    private Category category;

    @OneToMany(mappedBy="artwork")
    private List<NFT> nfts;
}

File: src/entity/artwork.ts
import {
    Column,
    Entity,
    ManyToOne,
    OneToMany,
    PrimaryGeneratedColumn,
} from 'typeorm';
import { Cart, Receipt, User } from '.';
import { Instruction } from '@metaplex-foundation/umi';

@Entity()
export class Artwork {
    @PrimaryGeneratedColumn()
    id: number;

    @Column()
    title: string;

    @Column()
    description: string;

    @Column()
    url: string;

    @Column()
    cid: string;

    @Column()
    metadata: string;

    @Column({ type: 'float' })
    cryptoPrice: number;

    @Column()
    currency: string;

    @ManyToOne(() => User, (user) => user.artworks)
    user: User;

    @OneToMany(() => Cart, (cart) => cart.artwork)
    carts: Cart[];

    @OneToMany(() => Receipt, (receipt) => receipt.artwork)
    receipts: Receipt[];

    @Column()
    published: boolean;

    @Column()
    createdAt: Date;

    @Column({ nullable: true, type: 'json' })
    instructions: Instruction[];

    @Column({ nullable: true })
    mint: string;
}
